sam/pileup.go

```diff
@@ -0,0 +1,366 @@
+	for read := range reads {
+		if !passesReadFilters(read, readFilters) {
+			continue
+		}
+		pb.sendPassed(read, includeNoData, refmap, send, pileFilters)
+		updatePile(pb, read, refmap)
+	}
+	close(send)
+}
+
+// passesReadFilters returns true if input Sam is true for all functions in filters.
+func passesReadFilters(s Sam, filters []func(s Sam) bool) bool {
+	for _, f := range filters {
+		if !f(s) {
+			return false
+		}
+	}
+	return true
+}
+
+// passesPileFilters returns true if input Sam is true for all functions in filters.
+func passesPileFilters(p Pile, filters []func(p Pile) bool) bool {
+	for _, f := range filters {
+		if !f(p) {
+			return false
+		}
+	}
+	return true
+}
+
+// updatePile updates the input pileBuffer with the data in buf
+func updatePile(pb *pileBuffer, s Sam, refmap map[string]chromInfo.ChromInfo) {
+	var seqPos int
+	refPos := s.Pos
+	refidx := refmap[s.RName].Order
+	for i := range s.Cigar {
+		switch s.Cigar[i].Op {
+		case 'M', '=', 'X': // Match
+			addMatch(pb, refidx, refPos, s.Seq[seqPos:seqPos+s.Cigar[i].RunLength])
+			refPos += uint32(s.Cigar[i].RunLength)
+			seqPos += s.Cigar[i].RunLength
+
+		case 'D': // Deletion
+			addDeletion(pb, refidx, refPos, s.Cigar[i].RunLength)
+			refPos += uint32(s.Cigar[i].RunLength)
+
+		case 'I': // Insertion
+			addInsertion(pb, refidx, refPos-1, s.Seq[seqPos:seqPos+s.Cigar[i].RunLength])
+			seqPos += s.Cigar[i].RunLength
+
+		default:
+			if cigar.ConsumesReference(s.Cigar[i].Op) {
+				refPos += uint32(s.Cigar[i].RunLength)
+			}
+			if cigar.ConsumesQuery(s.Cigar[i].Op) {
+				seqPos += s.Cigar[i].RunLength
+			}
+		}
+	}
+}
+
+// addMatch to pileBuffer
+func addMatch(pb *pileBuffer, refidx int, startPos uint32, seq []dna.Base) {
+	for i := range seq {
+		pb.addBase(refidx, startPos+uint32(i), seq[i])
+	}
+}
+
+// addDeletion to pileBuffer
+func addDeletion(pb *pileBuffer, refidx int, startPos uint32, length int) {
+	var i uint32
+	for i = 0; i < uint32(length); i++ {
+		pb.addBase(refidx, startPos+i, dna.Gap)
+	}
+}
+
+// addInsertion to pileBuffer
+func addInsertion(pb *pileBuffer, refidx int, startPos uint32, seq []dna.Base) {
+	pb.addIns(refidx, startPos, dna.BasesToString(seq))
+}
+
+// ****** pileBuffer functions below ****** //
+
+// pileBuffer stores a slice of piles for efficient reuse.
+type pileBuffer struct {
+	buf []Pile // discontinuous positions are not allowed
+	idx int    // index in buf of lowest pos
+}
+
+// newPileBuffer creates a newPileBuffer with input size
+func newPileBuffer(size int) *pileBuffer {
+	pb := pileBuffer{
+		buf: make([]Pile, size), // initialize to len of 2 standard reads
+	}
+	for i := range pb.buf {
+		pb.buf[i].RefIdx = -1
+		pb.buf[i].InsCount = make(map[string]int)
+	}
+	return &pb
+}
+
+// resetPile sets a Pile to the default state
+func resetPile(p *Pile) {
+	p.RefIdx = -1
+	p.Pos = 0
+	if !p.touched {
+		return
+	}
+	for i := range p.Count {
+		p.Count[i] = 0
+	}
+	p.InsCount = make(map[string]int)
```

Comment:
I had some trouble with this. Since the Piles are copied to the send channel, deleting the keys from the map could lead to concurrency errors as the map is passed by reference. However, you make a good point that the map will be empty most of the time, so I have transitioned to only making the map when it is needed, which should cut down on allocations.